nomad/state/state_store.go

```diff
@@ -4885,6 +4900,105 @@ func (s *StateStore) updateJobStabilityImpl(index uint64, namespace, jobID strin
 	return s.upsertJobImpl(index, nil, copy, true, txn)
 }
 
+func (s *StateStore) UpdateJobVersionTag(index uint64, namespace string, req *structs.JobApplyTagRequest) error {
+	jobID := req.JobID
+	jobVersion := req.Version
+	tag := req.Tag
+	name := req.Name
+
+	txn := s.db.WriteTxn(index)
+	defer txn.Abort()
+
+	// if no tag is present, this is a tag removal operation.
+	if tag == nil {
+		if err := s.unsetJobVersionTagImpl(index, namespace, jobID, name, txn); err != nil {
+			return err
+		}
+	} else {
+		if err := s.updateJobVersionTagImpl(index, namespace, jobID, jobVersion, tag, txn); err != nil {
+			return err
+		}
+	}
+
+	return txn.Commit()
+}
+
+func (s *StateStore) updateJobVersionTagImpl(index uint64, namespace, jobID string, jobVersion *uint64, tag *structs.JobTaggedVersion, txn *txn) error {
+	ws := memdb.NewWatchSet()
+
+	// Note: could use JobByIDAndVersion to get the specific version we want here,
+	// but then we'd have to make a second lookup to make sure we're not applying a duplicate tag name
+	versions, err := s.JobVersionsByID(ws, namespace, jobID)
+	if err != nil {
+		return err
+	}
+
+	duplicateVersionName := false
+	var job *structs.Job
+
+	for _, version := range versions {
+		// Allow for a tag to be updated (new description, for example) but otherwise don't allow a same-tagname to a different version.
+		if version.TaggedVersion != nil && version.TaggedVersion.Name == tag.Name && version.Version != *jobVersion {
+			duplicateVersionName = true
+			break
+		}
+		if version.Version == *jobVersion {
+			job = version
+		}
+	}
+
+	if duplicateVersionName {
+		return fmt.Errorf("tag %q already exists on a different version of job %q", tag.Name, jobID)
+	}
+
+	if job == nil {
+		return fmt.Errorf("job %q version %d not found", jobID, *jobVersion)
+	}
+
+	copy := job.Copy()
+	copy.TaggedVersion = tag
+	copy.ModifyIndex = index
+	return s.upsertJobVersion(index, copy, txn)
+}
+
+func (s *StateStore) UnsetJobVersionTag(index uint64, namespace, jobID string, name string) error {
+	txn := s.db.WriteTxn(index)
+	defer txn.Abort()
+
+	if err := s.unsetJobVersionTagImpl(index, namespace, jobID, name, txn); err != nil {
+		return err
+	}
+
+	return txn.Commit()
+}
+
+func (s *StateStore) unsetJobVersionTagImpl(index uint64, namespace, jobID string, name string, txn *txn) error {
+	ws := memdb.NewWatchSet()
+
+	versions, err := s.JobVersionsByID(ws, namespace, jobID)
+	if err != nil {
+		return err
+	}
+
+	var job *structs.Job
+
+	for _, version := range versions {
+		if version.TaggedVersion != nil && version.TaggedVersion.Name == name {
+			job = version
+			break
+		}
+	}
+
+	if job == nil {
+		return fmt.Errorf("tag %q not found on job %q", name, jobID)
+	}
```

Comment:
Isn't this `JobVersionByTagName`?

Reply:
It is, thanks for catching. I think I copied over the `updateTag` code here instead because it does duplicate-name checking (and this won't have to)